### PR TITLE
FIx SequenceReader Rewind(0) state corruption

### LINKED ISSUE #68774: SequenceReader<T>.Rewind(0) causes TryRead() to throw IndexOutOfRangeException
### Description

If `SequenceReader<T>.Rewind(0)` is called, any subsequent calls to `TryRead()` will throw IndexOutOfRange exception, as Rewind sets `_moreData` to false without checking if it actually rewound.

### Reproduction Steps

Here's a minimal program which exhibits the bug:
```cs
using System;
using System.Buffers;
using System.IO.Pipelines;
using System.Text;

var pipe = new Pipe();

pipe.Writer.Write(Encoding.UTF8.GetBytes("hello world"));
pipe.Writer.FlushAsync().AsTask().GetAwaiter().GetResult();
pipe.Writer.Complete();

if (!pipe.Reader.TryRead(out var result))
    Console.WriteLine("Read failed");

var reader = new SequenceReader<byte>(result.Buffer);

while (true)
{
    if (!reader.TryRead(out byte x))
        break;

    var offset = reader.Sequence.GetOffset(reader.Position);
    Console.WriteLine($"Success (ofs={offset}, x={(x == 10 ? "\\n" : (char)x)})");

    reader.Rewind(0); // cause bug
}
```

### Expected behavior

`Rewind(0)` should either fail (ArgumentOutOfRangeException?) or it should not set `_moreData` to false.

### Actual behavior

The SequenceReader's `_moreData` field gets set to `false`, meaning that a subsequent `TryRead` does an out-of-bounds read on `CurrentSpan`, throwing an IndexOutOfRangeException.

### Regression?

This seems to have been an issue as far back as 3.1.23, at least going through the runtimes I have installed

### Known Workarounds

Check that the amount to rewind is greater than zero before calling `Rewind`

### Configuration

The project is targeting `net6.0`.

```
$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.201
 Commit:    ef40e6aa06

Runtime Environment:
 OS Name:     debian
 OS Version:  
 OS Platform: Linux
 RID:         debian-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.201/

Host (useful for support):
  Version: 6.0.3
  Commit:  c24d9a9c91

.NET SDKs installed:
  3.1.417 [/usr/share/dotnet/sdk]
  5.0.406 [/usr/share/dotnet/sdk]
  6.0.100-preview.4.21255.9 [/usr/share/dotnet/sdk]
  6.0.100-preview.7.21379.14 [/usr/share/dotnet/sdk]
  6.0.201 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.23 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.15 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.4.21253.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.7.21378.6 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.23 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.15 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.4.21253.7 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.7.21377.19 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

### Other information

_No response_

### PR DESCRIPTION
**Pull request by hrrrrustic**
Close #68774

### CODE REVIEW

File: src/libraries/System.Memory/tests/SequenceReader/Rewind.cs [hunk 1]
```diff
@@ -120,5 +120,20 @@ public void RewindEmptyFirstSpan()
...
+            var emptySeq = ReadOnlySequence<byte>.Empty;
+            var reader = new SequenceReader<byte>(emptySeq);
+
+            Assert.True(reader.End);
+            Assert.False(reader.TryRead(out byte _));
+
+            reader.Rewind(0);
+
+            Assert.True(reader.End);
+            Assert.False(reader.TryRead(out byte _));
```
Comment (hrrrrustic) [inline src/libraries/System.Memory/tests/SequenceReader/Rewind.cs:136]: `Rewind(0)` set _moreData to true and all other methods thought that's still not the end

File: src/libraries/System.Memory/src/System/Buffers/SequenceReader.cs [hunk 2]
```diff
@@ -44,7 +44,7 @@ public SequenceReader(ReadOnlySequence<T> sequence)
         /// <summary>
         /// The underlying <see cref="ReadOnlySequence{T}"/> for the reader.
         /// </summary>
-        public readonly ReadOnlySequence<T> Sequence { get; }
+        public ReadOnlySequence<T> Sequence { get; }
```
Comment (teo-tsirpanis) [inline src/libraries/System.Memory/src/System/Buffers/SequenceReader.cs:47]: Why did you remove the `readonly`?

Comment (hrrrrustic) [reply src/libraries/System.Memory/src/System/Buffers/SequenceReader.cs:47]: Auto-properties are already readonly

Comment (teo-tsirpanis) [reply src/libraries/System.Memory/src/System/Buffers/SequenceReader.cs:47]: TIL. Thanks.

File: src/libraries/System.Memory/src/System/Buffers/SequenceReader.cs [hunk 3]
```diff
@@ -223,6 +223,9 @@ public void Rewind(long count)
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count);
             }
 
+            if (count == 0)
+                return;
```
Comment (adamsitnik) [inline src/libraries/System.Memory/src/System/Buffers/SequenceReader.cs:227]: nit: we need braces even for one liners to be consistent with other if blocks in this method

```suggestion
            if (count == 0)
            {
                return;
            }
```

**Review by adamsitnik — APPROVED** [review]
LGTM, I believe you made the right decision.

Thank you for your contribution @hrrrrustic !
